format/ComputeIR.fbs

```diff
@@ -0,0 +1,521 @@
+  having:[ArrayExpr];
+}
+
+/// Select up to the indicated number of rows from the input expression based
+/// on the first-emitted rows when evaluating the input expression. Generally,
+/// no particular order is guaranteed unless combined with a sort expression.
+table Limit {
+  input:TableExpr;
+
+  /// Number of logical rows to select from input.
+  limit:long;
+}
+
+enum RelationalJoinType : int {
+  INNER = 0,
+  LEFT = 1,
+  RIGHT = 2,
+  OUTER = 3,
+  SEMI = 4,
+  ANTI = 5,
+}
+
+/// A standard relational / SQL-style equijoin.
+///
+/// Providing no left/right columns produces the cross product of the two
+/// tables.
+table EqualityJoin {
+  // TODO: complete and document
+  type:RelationalJoinType = INNER;
+
+  left:TableExpr (required);
+  right:TableExpr (required);
+
+  /// The columns from the left expression to use when joining
+  left_columns:[ColumnReference] (required);
+
+  /// The columns from the right expression to use when joining. Must have the
+  /// same length as left_columns.
+  ///
+  /// If omitted, the names provided in left_columns must match the names in
+  /// the right expression.
+  right_columns:[ColumnReference];
+}
+
+/// A relational non-equijoin containing expressions which may include
+/// inequality or range conditions.
+table NonEqualityJoin {
+  // TODO: complete and document
+  type:RelationalJoinType = INNER;
+  left:TableExpr (required);
+  right:TableExpr (required);
+  left_exprs:[ArrayExpr];
+  right_exprs:[ArrayExpr];
+}
+
+/// A temporal join type
+///
+/// TODO: complete and document
+table AsOfJoin {
+  left:TableExpr (required);
+  right:TableExpr;
+
+  /// The column in the left expression to use for data ordering to determine
+  /// the "as of" time.
+  left_asof:ColumnReference (required);
+
+  /// The column in the right expression to use for data ordering to determine
+  /// the "as of" time. If the column name is the same as left_asof, may be
+  /// omitted.
+  right_asof:ColumnReference;
+
+  /// TODO: Define means of providing time deltas in this IR.
+  tolerance:Literal;
+
+  /// If true, the
+  allow_equal:bool = true;
+}
+
+/// An extension of as-of join which allows applying an aggregate function to
+/// the data falling within the indicated time interval.
+///
+/// TODO: Define semantics of "identity" window where all elements of window
+/// become a List<T> element in the result.
+table WindowJoin {
+  // TODO
+}
+
+/// An expression to use for sorting. The key expression determines the values
+/// to be used for ordering the table's rows.
+table SortKey {
+  key:ArrayExpr (required);
+  ascending:bool = true;
+}
+
+/// A (possibly hierarchical) sort operation with one or more sort keys.
+table Sort {
+  // TODO
+  keys:[SortKey] (required);
+}
+
+/// A table-generating function.
+///
+/// TODO
+table TableFunction {
+  descr:FunctionDescr (required);
+
+  /// An optional output schema for the table function. If not provided, must
+  /// be determined by the IR consumer.
```

Comment:
What would a consumer do given a missing `out_schema`? I think there's no ambiguity if we stipulate that the consumer must conform to the output type.